src/charm.py

```diff
@@ -1,19 +1,497 @@
+            if paths[0].user != Config.UNIX_USER or paths[0].group != Config.UNIX_GROUP:
+                container.exec(
+                    f"chown {Config.UNIX_USER}:{Config.UNIX_GROUP} -R {path}".split(" ")
+                )
+
+    def push_file_to_unit(
+        self,
+        parent_dir: str,
+        file_name: str,
+        file_contents: str,
+        container: Container = None,
+    ) -> None:
+        """Push the file on the container, with the right permissions."""
+        container = container or self.unit.get_container(Config.CONTAINER_NAME)
+        container.push(
+            f"{parent_dir}/{file_name}",
+            file_contents,
+            make_dirs=True,
+            permissions=0o400,
+            user=Config.UNIX_USER,
+            group=Config.UNIX_GROUP,
+        )
+
+    def unit_host(self, unit: Unit) -> str:
+        """Create a DNS name for a MongoDB unit.
+
+        Args:
+            unit_name: the juju unit name, e.g. "mongodb/1".
+
+        Returns:
+            A string representing the hostname of the MongoDB unit.
+        """
+        unit_id = unit.name.split("/")[1]
+        return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
+
+    # END: helper functions
+
+    # BEGIN: properties
+    @property
+    def mongos_initialised(self) -> bool:
+        """Check if mongos is initialised."""
+        return "mongos_initialised" in self.app_peer_data
+
+    @mongos_initialised.setter
+    def mongos_initialised(self, value: bool):
+        """Set the mongos_initialised flag."""
+        if value:
+            self.app_peer_data["mongos_initialised"] = str(value)
+        elif "mongos_initialised" in self.app_peer_data:
+            del self.app_peer_data["mongos_initialised"]
+
+    @property
+    def _unit_ip(self) -> str:
+        """Returns the ip address of the unit."""
+        return str(self.model.get_binding(Config.Relations.PEERS).network.bind_address)
+
+    @property
+    def is_external_client(self) -> Optional[str]:
+        """Returns the connectivity mode which mongos should use.
+
+        This is determined by checking the modes requested by the client(s).
+
+        TODO: Future PR. This should be modified to work for many clients.
+        """
+        if EXTERNAL_CONNECTIVITY_TAG not in self.app_peer_data:
+            return False
+
+        return json.loads(self.app_peer_data.get(EXTERNAL_CONNECTIVITY_TAG))
+
+    @property
+    def database(self) -> Optional[str]:
+        """Returns a mapping of databases requested by integrated clients.
+
+        TODO: Future PR. This should be modified to work for many clients.
+        """
+        if not self._peers:
+            logger.info("Peer relation not joined yet.")
+            # TODO future PR implement relation interface between host application mongos and use
+            # host application name in generation of db name.
+            return "mongos-database"
```

Comment:
I think we should either raise an exception or return None

Reply:
this function is already covered in the subsequent PR which is currently a WIP. I would prefer to not make this change in this PR 

If this request is blocking please let me know and I can make that change